File: passform2_ws/src/passform_ros-main/passform_util/passform_util/types/master.py
import uuid

from basyx.aas import model
from basyx.aas.model import base
from passform_util.basyx import sanitize_basyx_string
from passform_util.types import Bay


class MasterAsset(model.Asset):
    def __init__(self,
        unique_id: str
        ):

        super().__init__(
            kind = model.base.AssetKind.INSTANCE,
            identification=model.Identifier(unique_id+'_ASSET', model.IdentifierType.CUSTOM),
            category='passform_master_asset'
        )

class TechnicalData(model.Submodel):
    def __init__(
        self,
        unique_id: str
    ):
        super().__init__(
            identification=model.Identifier(unique_id+'_TD', model.IdentifierType.CUSTOM),
            id_short='TechnicalData',
        )

class Master(model.AssetAdministrationShell):
    def __init__(self,
        unique_id: str = str(uuid.uuid4()),
        name: str = 'PassForM_Master',
        # properties: List[model.Property] = field(default_factory=list, compare=False),
        **kwargs):

        super().__init__(
            identification=model.Identifier(unique_id, model.IdentifierType.CUSTOM),
            id_short=sanitize_basyx_string(name), #replace ' ', '-'
            category='passform_master',
            submodel={TechnicalData(unique_id)},
            asset=MasterAsset(unique_id),
            **kwargs)

    def create_bay(self, bay_id, origin=[0.0,0.0,0.0]):
        b = Bay(str(uuid.uuid4()), 'Bay'+str(bay_id), origin=origin)
        self.submodel.add(b)

    def get_bays(self):
        resp = list()
        for s in self.submodel:
            if s.category == "passform_bay":
                resp.append(s)
        return resp

if __name__ == '__main__':
    import json
    import unittest

    import basyx.aas.adapter.json as basyx_json
    from passform_util.basyx.rest import RestServer

    m = Master('uuid')
    m.create_bay(1)
    m.create_bay(2)
    for b in m.get_bays():
        print(b)
    # data_in = json.loads(json.dumps(m, cls=basyx_json.AASToJsonEncoder))
    #
    # api = RestServer()
    # api.remove_aas('uuid')
    #
    # api.add_aas(m)
    # m2 = api.get_aas('uuid')
    # data_out = json.loads(json.dumps(m2, cls=basyx_json.AASToJsonEncoder))
    #
    # tc = unittest.TestCase()
    # tc.assertDictEqual(data_in, data_out)

File: passform2_agent_backend/src/core/types/master.rs
use serde::{Serialize, Deserialize};
use serde_json::{json, Value};
use uuid::Uuid;
use crate::core::util::util::sanitize_id;
// Korrigierter Pfad: BaySubmodel liegt in crate::core::types::bay
use crate::core::types::bay::BaySubmodel; 

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Master {
    pub unique_id: String,
    pub name: String,
    pub bays: Vec<BaySubmodel>,
}

impl Master {
    pub fn new(unique_id: Option<String>, name: Option<String>) -> Self {
        Self {
            unique_id: unique_id.unwrap_or_else(|| Uuid::new_v4().to_string()),
            // sanitize_id erwartet einen &str, das passt hier
            name: sanitize_id(&name.unwrap_or_else(|| "PassForM_Master".to_string())),
            bays: Vec::new(),
        }
    }

    /// Erstellt eine neue Bay und fügt sie dem Master hinzu
    pub fn create_bay(&mut self, bay_id: &str, origin: [f64; 3]) {
        let bay_uuid = Uuid::new_v4().to_string();
        let bay_name = format!("Bay{}", bay_id);
        let new_bay = BaySubmodel::new(&bay_uuid, &bay_name, origin, false);
        self.bays.push(new_bay);
    }

    /// Gibt alle Submodelle vom Typ "passform_bay" zurück
    pub fn get_bays(&self) -> &Vec<BaySubmodel> {
        &self.bays
    }

    /// Generiert die komplette AAS-Struktur als JSON für den BaSyx-Server
    pub fn to_basyx_json(&self) -> Value {
        // Fix für E0282: Explizite Typangabe für den Closure-Parameter 'b'
        let submodels: Vec<Value> = self.bays
            .iter()
            .map(|b: &BaySubmodel| b.to_basyx_json()) 
            .collect();

        // Füge das TechnicalData Submodell des Masters hinzu
        let mut all_submodels = submodels;
        all_submodels.push(json!({
            "idShort": "TechnicalData",
            "identification": {
                "id": format!("{}_TD", self.unique_id),
                "idType": "CUSTOM"
            },
            "modelType": "Submodel",
            "submodelElements": []
        }));

        json!({
            "idShort": self.name,
            "identification": {
                "id": self.unique_id,
                "idType": "CUSTOM"
            },
            "category": "passform_master",
            "modelType": "AssetAdministrationShell",
            "asset": {
                "idShort": "MasterAsset",
                "identification": {
                    "id": format!("{}_ASSET", self.unique_id),
                    "idType": "CUSTOM"
                },
                "assetKind": "INSTANCE"
            },
            "submodels": all_submodels
        })
    }
}